### PR TITLE
[dhcp_relay]: Check whether device requires DHCP relay upon docker start

### PR DESCRIPTION
**Pull request by jleveque**
If not, log a message and exit docker. Other conditions can be added in the future as needed.

With OneImage, all dockers must be deployed, but DHCP relay may not be needed. It's cleaner to have docker exit than to continue running and serving no purpose. Log message is helpful to understand why.

Signed-off-by: Joe LeVeque <jolevequ@microsoft.com>

### CODE REVIEW

File: dockers/docker-dhcp-relay/config.sh [hunk 1]
```diff
@@ -1,4 +1,12 @@
 #!/bin/bash
 
-sonic-cfggen -m /etc/sonic/minigraph.xml -y /etc/sonic/dhcp_relay.yml -t /usr/share/sonic/templates/isc-dhcp-relay.j2 > /etc/default/isc-dhcp-relay
+MINIGRAPH_HOSTNAME=`sonic-cfggen -m /etc/sonic/minigraph.xml -v "minigraph_hostname"`
+DEVICE_ROLE=`sonic-cfggen -m /etc/sonic/minigraph.xml -v "minigraph_devices['$MINIGRAPH_HOSTNAME']['type']"`
+
+if [ $DEVICE_ROLE != "ToRRouter" ]; then
+    echo "Device does not require DHCP relay. docker-dhcp-relay exiting..."
```
Comment (stcheng) [inline dockers/docker-dhcp-relay/config.sh]: if the script exits here, will the docker exits?

Comment (taoyl-ms) [inline dockers/docker-dhcp-relay/config.sh]: I think if you return a nonzero value, it will.

File: dockers/docker-dhcp-relay/config.sh [hunk 2]
```diff
@@ -1,4 +1,12 @@
 #!/bin/bash
 
-sonic-cfggen -m /etc/sonic/minigraph.xml -y /etc/sonic/dhcp_relay.yml -t /usr/share/sonic/templates/isc-dhcp-relay.j2 > /etc/default/isc-dhcp-relay
+MINIGRAPH_HOSTNAME=`sonic-cfggen -m /etc/sonic/minigraph.xml -v "minigraph_hostname"`
+DEVICE_ROLE=`sonic-cfggen -m /etc/sonic/minigraph.xml -v "minigraph_devices['$MINIGRAPH_HOSTNAME']['type']"`
+
```
Comment (taoyl-ms) [inline dockers/docker-dhcp-relay/config.sh]: you can directly use -v "minigraph_devices[minigraph_hostname]['type']", don't need to call it twice.

Comment (jleveque) [inline dockers/docker-dhcp-relay/config.sh]: Good to know. Simplified in https://github.com/Azure/sonic-buildimage/pull/371/commits/265311e7af283ec8c62bb950ccfc348dcfa787d2. Thanks!

File: dockers/docker-dhcp-relay/config.sh (hunk 1 above)
Comment (jleveque) [inline dockers/docker-dhcp-relay/config.sh]: Taoyu is correct. I have tested the behavior, and exiting with a nonzero value will cause the docker container to exit, whereas exiting with a return value of 0 will cause the container to keep running.

**Comment by lguohan** [comment]
should we do this in dhcprelay service unit file? if we detect it is not tor, we do not even start the container.

**Comment by jleveque** [comment]
It would be cleaner to have this logic outside the container, but systemd provides a [limited set of conditionals](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#ConditionArchitecture=) to work with in the unit files, none of which I'm aware of allow us to directly check the device type from the minigraph. Do you have an idea in mind?

**Comment by taoyl-ms** [comment]
How about doing it in /usr/bin/dhcp_relay.sh instead of in the service definition file?

**Comment by jleveque** [comment]
Thanks for the idea, but after speaking with Guohan and Shuotian, we decided the service shouldn't even be started if not needed. This also applies to teamd, and may apply to other services in the future. I will soon submit another solution that disables the services on first boot.

**Comment by taoyl-ms** [comment]
If you are going to disable the service on first boot, please make sure your disabling happens after update_graph service so we don't mistakenly disable it based on a default graph.

**Comment by jleveque** [comment]
Thanks for the heads-up! :)

**Comment by jleveque** [comment]
Abandoning in favor of https://github.com/Azure/sonic-buildimage/pull/427

**Pull request closed without merging**